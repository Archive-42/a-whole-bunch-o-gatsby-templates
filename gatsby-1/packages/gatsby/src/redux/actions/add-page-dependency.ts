import { store } from "../"

import { createPageDependency as internalCreatePageDependency } from "./internal"

export const createPageDependency = ({
  path,
  nodeId,
  connection,
}: {
  path: string
  nodeId: string
  connection?: string
}): void => {
  const { queries } = store.getState()

  // Check that the dependencies aren't already recorded so we
  // can avoid creating lots of very noisy actions.
  let nodeDependencyExists = false
  let connectionDependencyExists = false
  if (!nodeId) {
    nodeDependencyExists = true
  }
  if (
    nodeId &&
    queries.byNode.has(nodeId) &&
    queries.byNode.get(nodeId)!.has(path)
  ) {
    nodeDependencyExists = true
  }
  if (!connection) {
    connectionDependencyExists = true
  }
  if (
    connection &&
    queries.byConnection.has(connection) &&
    queries.byConnection.get(connection)!.has(path)
  ) {
    connectionDependencyExists = true
  }
  if (nodeDependencyExists && connectionDependencyExists) {
    return
  }

  // It's new, let's dispatch it
  store.dispatch(internalCreatePageDependency({ path, nodeId, connection }))
}
