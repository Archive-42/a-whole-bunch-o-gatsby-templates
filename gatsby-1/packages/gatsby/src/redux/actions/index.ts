import * as internalActions from "./internal"
import { actions as publicActions } from "./public"
import {
  actions as restrictedActions,
  availableActionsByAPI,
} from "./restricted"

export const actions = {
  ...internalActions,
  ...publicActions,
  ...restrictedActions,
}

export const restrictedActionsAvailableInAPI = availableActionsByAPI

export { internalActions, publicActions, restrictedActions }
