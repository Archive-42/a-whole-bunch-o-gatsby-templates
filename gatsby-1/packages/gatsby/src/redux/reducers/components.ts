import normalize from "./normalize-path"
import { IGatsbyState, ActionsUnion } from "../types"

let programStatus = `BOOTSTRAPPING`

// TODO: replace usages of this reducer with queries.trackedComponents
//  It is here merely for compatibility.
export const componentsReducer = (
  state: IGatsbyState["components"] = new Map(),
  action: ActionsUnion
): IGatsbyState["components"] => {
  switch (action.type) {
    case `DELETE_CACHE`:
      return new Map()
    case `SET_PROGRAM_STATUS`:
      programStatus = action.payload
      return state
    case `CREATE_PAGE`: {
      action.payload.componentPath = normalize(action.payload.component)
      // Create XState service.
      let component = state.get(action.payload.componentPath)
      if (!component) {
        component = {
          componentPath: action.payload.componentPath,
          componentChunkName: action.payload.componentChunkName,
          query: ``,
          pages: new Set(),
          isInBootstrap: true,
        }
      }
      component.pages.add(action.payload.path)
      component.isInBootstrap = programStatus === `BOOTSTRAPPING`
      state.set(action.payload.componentPath, component)
      return state
    }
    case `QUERY_EXTRACTED`: {
      action.payload.componentPath = normalize(action.payload.componentPath)
      const component = state.get(action.payload.componentPath)!
      component.query = action.payload.query
      state.set(action.payload.componentPath, component)
      return state
    }
    case `REMOVE_STATIC_QUERIES_BY_TEMPLATE`: {
      action.payload.componentPath = normalize(action.payload.componentPath)
      state.delete(action.payload.componentPath)
      return state
    }
    case `DELETE_PAGE`: {
      const component = state.get(normalize(action.payload.component))!
      component.pages.delete(action.payload.path)
      return state
    }
  }

  return state
}
