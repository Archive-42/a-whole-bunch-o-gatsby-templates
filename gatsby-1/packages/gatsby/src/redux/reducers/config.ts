import { IGatsbyConfig, ISetSiteConfig } from "../types"

export const configReducer = (
  state: IGatsbyConfig = {},
  action: ISetSiteConfig
): IGatsbyConfig => {
  switch (action.type) {
    case `SET_SITE_CONFIG`: {
      return action.payload
    }
    default:
      return state
  }
}
