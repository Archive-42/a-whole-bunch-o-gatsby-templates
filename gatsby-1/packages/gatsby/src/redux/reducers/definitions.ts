import { ActionsUnion, IGatsbyState } from "../types"

export const definitionsReducer = (
  state: IGatsbyState["definitions"] = new Map(),
  action: ActionsUnion
): IGatsbyState["definitions"] => {
  switch (action.type) {
    case `SET_GRAPHQL_DEFINITIONS`:
      return action.payload
    default:
      return state
  }
}
