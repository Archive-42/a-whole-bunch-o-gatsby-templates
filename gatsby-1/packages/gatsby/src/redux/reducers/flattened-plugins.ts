import { IGatsbyState, ActionsUnion } from "../types"

export const flattenedPluginsReducer = (
  state: IGatsbyState["flattenedPlugins"] = [],
  action: ActionsUnion
): IGatsbyState["flattenedPlugins"] => {
  switch (action.type) {
    case `SET_SITE_FLATTENED_PLUGINS`:
      return [...action.payload]

    default:
      return state
  }
}
