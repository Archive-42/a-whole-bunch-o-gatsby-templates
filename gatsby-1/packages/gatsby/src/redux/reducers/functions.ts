import { IGatsbyState, ISetSiteFunctions } from "../types"

export const functionsReducer = (
  state: IGatsbyState["functions"] = [],
  action: ISetSiteFunctions
): IGatsbyState["functions"] => {
  switch (action.type) {
    case `SET_SITE_FUNCTIONS`:
      return action.payload

    default:
      return state
  }
}
