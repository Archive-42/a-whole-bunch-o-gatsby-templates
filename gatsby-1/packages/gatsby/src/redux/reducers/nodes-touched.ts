import { ActionsUnion } from "./../types"
import { IGatsbyState } from "gatsby/src/redux/types"

export const nodesTouchedReducer = (
  state: IGatsbyState["nodesTouched"] = new Set(),
  action: ActionsUnion
): IGatsbyState["nodesTouched"] => {
  switch (action.type) {
    case `CREATE_NODE`:
      state.add(action.payload.id)
      return state

    case `TOUCH_NODE`:
      state.add(action.payload)
      return state

    default:
      return state
  }
}
