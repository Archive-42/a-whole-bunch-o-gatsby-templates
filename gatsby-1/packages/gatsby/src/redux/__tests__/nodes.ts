import { actions } from "../actions"
import { nodesReducer } from "../reducers/nodes"
import { IGatsbyNode } from "../types"
import { nodesTouchedReducer } from "../reducers/nodes-touched"

const dispatch = jest.fn()

type MapObject = Record<string, IGatsbyNode>

const fromMapToObject = (map: Map<string, any>): MapObject => {
  const obj: MapObject = {}
  Array.from(map.entries()).forEach(([key, value]) => {
    obj[key] = value
  })
  return obj
}

describe(`Create and update nodes`, (): void => {
  beforeEach((): void => {
    dispatch.mockClear()
  })

  it(`allows creating nodes`, (): void => {
    actions.createNode(
      {
        id: `hi`,
        children: [],
        parent: `test`,
        internal: {
          contentDigest: `hasdfljds`,
          type: `Test`,
        },
        pickle: true,
      },
      {
        name: `tests`,
      }
    )(dispatch)
    const action = dispatch.mock.calls[0][0]
    expect(action).toMatchSnapshot({
      payload: { internal: { counter: expect.any(Number) } },
    })
    expect(fromMapToObject(nodesReducer(undefined, action))).toMatchSnapshot({
      hi: { internal: { counter: expect.any(Number) } },
    })
  })

  it(`allows updating nodes`, (): void => {
    actions.createNode(
      {
        id: `hi`,
        children: [],
        parent: `test`,
        internal: {
          contentDigest: `hasdfljds`,
          type: `Test`,
        },
        pickle: true,
        deep: {
          array: [
            0,
            1,
            {
              boom: true,
            },
          ],
        },
      },
      {
        name: `tests`,
      }
    )(dispatch)
    const action = dispatch.mock.calls[0][0]

    actions.createNode(
      {
        id: `hi`,
        children: [],
        parent: `test`,
        internal: {
          contentDigest: `hasdfljds`,
          type: `Test`,
        },
        pickle: false,
        deep: {
          array: [1, 2],
        },
        deep2: {
          boom: `foo`,
        },
      },
      {
        name: `tests`,
      }
    )(dispatch)
    const updateAction = dispatch.mock.calls[1][0]

    let state = nodesReducer(undefined, action)
    state = nodesReducer(state, updateAction)

    expect(state.get(`hi`)!.pickle).toEqual(false)
    expect((state.get(`hi`)!.deep as any).array![0]).toEqual(1)
    expect((state.get(`hi`)!.deep2 as any).boom).toEqual(`foo`)
  })

  it(`nodes that are added are also "touched"`, (): void => {
    actions.createNode(
      {
        id: `hi`,
        children: [],
        parent: `test`,
        internal: {
          contentDigest: `hasdfljds`,
          type: `Test`,
        },
        pickle: true,
      },
      {
        name: `tests`,
      }
    )(dispatch)
    const action = dispatch.mock.calls[0][0]

    const state = nodesTouchedReducer(undefined, action)

    expect(state instanceof Set).toBe(true)

    expect(state.has(`hi`)).toBe(true)
  })

  it(`allows adding fields to nodes`, (): void => {
    actions.createNode(
      {
        id: `hi`,
        children: [],
        parent: `test`,
        internal: {
          contentDigest: `hasdfljds`,
          type: `Test`,
        },
        pickle: true,
      },
      {
        name: `tests`,
      }
    )(dispatch)
    const action = dispatch.mock.calls[0][0]
    let state = nodesReducer(undefined, action)

    const addFieldAction = actions.createNodeField(
      {
        node: state.get(`hi`),
        name: `joy`,
        value: `soul's delight`,
      },
      {
        name: `test`,
      }
    )

    state = nodesReducer(state, addFieldAction)
    expect(fromMapToObject(state)).toMatchSnapshot({
      hi: { internal: { counter: expect.any(Number) } },
    })
  })

  it(`throws error if a field is updated by a plugin not its owner`, (): void => {
    actions.createNode(
      {
        id: `hi`,
        children: [],
        parent: `test`,
        internal: {
          contentDigest: `hasdfljds`,
          type: `Test`,
        },
        pickle: true,
      },
      {
        name: `tests`,
      }
    )(dispatch)
    const action = dispatch.mock.calls[0][0]
    let state = nodesReducer(undefined, action)

    const addFieldAction = actions.createNodeField(
      {
        node: state.get(`hi`),
        name: `joy`,
        value: `soul's delight`,
      },
      {
        name: `test`,
      }
    )
    state = nodesReducer(state, addFieldAction)

    function callActionCreator(): void {
      actions.createNodeField(
        {
          node: state.get(`hi`),
          name: `joy`,
          value: `soul's delight`,
        },
        {
          name: `test2`,
        }
      )
    }
    expect(callActionCreator).toThrowError(
      `A plugin tried to update a node field that it doesn't own`
    )
  })

  it(`throws error if a node is created by a plugin not its owner`, (): void => {
    actions.createNode(
      {
        id: `hi`,
        children: [],
        parent: `test`,
        internal: {
          contentDigest: `hasdfljds`,
          type: `mineOnly`,
        },
        pickle: true,
      },
      {
        name: `pluginA`,
      }
    )(dispatch)

    function callActionCreator(): void {
      actions.createNode(
        {
          id: `hi2`,
          children: [],
          parent: `test`,
          internal: {
            contentDigest: `hasdfljds`,
            type: `mineOnly`,
          },
          pickle: true,
        },
        {
          name: `pluginB`,
        }
      )(dispatch)
    }

    expect(callActionCreator).toThrowError(
      `The plugin "pluginB" created a node of a type owned by another plugin.`
    )
  })

  it(`throws error if a node sets a value on "fields"`, (): void => {
    function callActionCreator(): void {
      actions.createNode(
        {
          id: `hi`,
          children: [],
          parent: `test`,
          fields: {
            test: `I can't do this but I like to test boundaries`,
          },
          internal: {
            contentDigest: `hasdfljds`,
            type: `mineOnly`,
          },
          pickle: true,
        },
        {
          name: `pluginA`,
        }
      )(dispatch)
    }

    expect(callActionCreator).toThrowError(
      `Plugins creating nodes can not set data on the reserved field "fields"
      as this is reserved for plugins which wish to extend your nodes.`
    )
  })
})
