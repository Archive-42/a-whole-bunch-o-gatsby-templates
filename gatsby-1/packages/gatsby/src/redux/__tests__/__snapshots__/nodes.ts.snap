// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create and update nodes allows adding fields to nodes 1`] = `
Object {
  "hi": Object {
    "children": Array [],
    "fields": Object {
      "joy": "soul's delight",
    },
    "id": "hi",
    "internal": Object {
      "contentDigest": "hasdfljds",
      "counter": Any<Number>,
      "fieldOwners": Object {
        "joy": "test",
      },
      "owner": "tests",
      "type": "Test",
    },
    "parent": "test",
    "pickle": true,
  },
}
`;

exports[`Create and update nodes allows creating nodes 1`] = `
Object {
  "oldNode": undefined,
  "payload": Object {
    "children": Array [],
    "id": "hi",
    "internal": Object {
      "contentDigest": "hasdfljds",
      "counter": Any<Number>,
      "owner": "tests",
      "type": "Test",
    },
    "parent": "test",
    "pickle": true,
  },
  "plugin": Object {
    "name": "tests",
  },
  "type": "CREATE_NODE",
}
`;

exports[`Create and update nodes allows creating nodes 2`] = `
Object {
  "hi": Object {
    "children": Array [],
    "id": "hi",
    "internal": Object {
      "contentDigest": "hasdfljds",
      "counter": Any<Number>,
      "owner": "tests",
      "type": "Test",
    },
    "parent": "test",
    "pickle": true,
  },
}
`;
